### ISSUE TITLE
Session invalidation creates extra (unused) sessions

### ISSUE THREAD

**Issue by Deimos** [jvanasco/pyramid_session_redis#2]
Hi @jvanasco,

I came across your fork while looking into pyramid_redis_sessions and wanted to try it out since it seems like you've made some good updates. However, I noticed a bug with session invalidation, which I think might be related to a hook that's inadvertently creating a new session after invalidating the previous one.

If I just create a very simple view that does nothing except `request.session.invalidate()`, it will end up creating two sessions each time I load it, with the second session being stored in redis but not used. This behavior doesn't happen in pyramid_redis_sessions with an identical setup.

For a specific example, starting with empty redis:

```
127.0.0.1:6379> keys *
(empty list or set)
```

Now if I load the view that just invalidates a session, the session id invalidated is `nSHEXxd0mqa72Q_SbidPdred-VAslmbboT11aidFx8lHLrkAL5FVjyT1ROmuQA3S`, but checking redis now, there's a second session with a different ID present:

```
127.0.0.1:6379> keys *
1) "session:jBxmZV2vnddM41Tj1YLIkwjChDotFtHgClqztYrMmbkjCSj5nEBC-OZVhG8NytL_"
```

Reloading the page will create and invalidate a new session (not the one with ID `jBxmZ...`), and add another to redis:

```
127.0.0.1:6379> keys *
1) "session:w-QNjKAjUJjTjFGcm-f68cCnQ9plOKGnbmYaWplUMbat8ZDpOeDJjjTx7LebSyQ0"
2) "session:jBxmZV2vnddM41Tj1YLIkwjChDotFtHgClqztYrMmbkjCSj5nEBC-OZVhG8NytL_"
```

Each additional load will continue this, with an extraneous session created each time.

**Comment by jvanasco** [comment]
ugh.  thanks for catching this.

i have a few ideas on how to fix it, and will get on it ASAP!

**Comment by jvanasco** [comment]
I was wrong... this is a weird edge case and will take a few days to fix.

FWIW, this is a pretty limited edge-case.  it only happens under two scenarios (which might just be one scenario, but I'll test against both)

   def case_1(request):
      	"""view only invalidates"""
      	request.session.invalidate()

   def case_2(request):
      	"""view invalidates after an update"""
      	request.session['a'] = request.session['a'] + 1 if 'a' in request.session else 0
      	request.session.invalidate()

the issue is that the session's internals are triggering a new session_id generation (it's not actually a session in redis, but a placeholder), however a cookie isn't firing out to the user.    on the second view, the client has no cookie... so it's repeating the same thing nonstop.

if you're worried about redis resources until this is fixed...

1. don't be. this is just a small placeholder and will expire out if you have a timeout set.
2. accessing the session in any way -- even reading `request.session.id` -- will trigger cookie persistence.

the fix i'm aiming for is to not create a session in redis on expiry; and just defer everything until the session is accessed again.  that will be a lot easier on system resources.  if i can't figure that out within a few days, I'll put that on a todo list and push an update that just creates a new cookie with the placeholder value.

**Comment by Deimos** [comment]
Thanks for looking into it so quickly. Like you said, I don't think it's harmful so I'm not particularly worried about it, just wanted to let you know since it seemed unintentional.

**Comment by jvanasco** [comment]
I had time to investigate this morning, and Fixed in "v1.2.1" (now on pypi).  We've got this running on a cluster and have had redis resourcing issues in the past, so any potential performance issues are a big concern.


The session's state object is deleted during `invalidate()` and checking the state object for refresh/persist information in the finished callback will generate a new state object and corresponding session placeholder... however it won't (can't?) register the set_cookie.  The fix was simply checking for the existence of a state object before potentially creating one.

Just a 2 line edit... https://github.com/jvanasco/pyramid_session_redis/commit/6b6867e4163f45982721fd27f671b2668500a22b?diff=unified#diff-91db0802223be236df29b719248197afR435
